lib/commit-lint.ts

```diff
@@ -0,0 +1,152 @@
+           if (!await this.lint3(lines, msgs)) result = false;
+           if (!await this.lint4(lines, msgs)) result = false;
+
+            // tslint:enable:curly
+        } catch (error) {
+            msgs.unshift(`Commit message checking stopped due to:\n${error}`);
+            result = false;
+        }
+
+        if (msgs.length) {
+            await this.addComment(`There ${
+                    msgs.length > 1 ?
+                    "are issues" : "is an issue"
+                    } in commit ${
+                patch.commit}:\n${msgs.join("\n")}`);
+        }
+
+        return result;
+    }
+/*
+TODO:
+There is also the convention that the commit message should talk in the
+imperative tone (as opposed to the past tense, as many other projects prefer).
+I am not sure that there is an easy way to test for that.
+*/
+    // This method tests:
+    // the body of the commit message should not be empty
+    // the first line should not exceed 76 characters
+    // the first line should be followed by an empty line
+
+    private async lint1(lines: string[], msgs: string[]):
+        Promise<boolean> {
+        let result: boolean = true;
+
+        if (lines.length < 3) {
+            throw new Error("Commit message is too short");
+        }
+
+        if (lines[0].length > 76) {
+            msgs.push("First line of commit message is too long " +
+                `(> 76 columns): ${lines[0]}`);
+            result = false;
+        }
+
+        if (lines[1].length !== 0) {
+            // tslint:disable-next-line:max-line-length
+            msgs.push("The first line must be separated from the rest by an empty line");
+            result = false;
+        }
+
+        return result;
+    }
+
+    // This method tests:
+    // if the first line starts with a prefix (e.g. tests:), it should continue
+    // in lower-case
+
+    private async lint2(lines: string[], msgs: string[]):
+        Promise<boolean> {
+        let result: boolean = true;
+
+        const match = lines[0].match(/^([a-z]+)+?:.*?([A-Z])/);
+
+        if (match && typeof(match[2]) === "string" ) {
+            msgs.push(`Prefixed commit message must be in lower case: ${
+                lines[0]}`);
+            result = false;
+        }
+
+        return result;
+    }
+
+    // This method tests:
+    // the last line should be a Signed-off-by: line - DCO check does this
+    // already, but put out a message if it is indented
+
+    private async lint3(lines: string[], msgs: string[]):
+        Promise<boolean> {
+        let result: boolean = true;
+        const line = lines[lines.length - 1];
+        const match = line.match(/^(\s*)Signed-off-by:\s+(.*)/);
+
+        if (match) {
+            if (match[1].length) {
+                msgs.push(`Leading whitespace in sign off: ${
+                    line}`);
+                result = false;
+            }
+        } else {
+            msgs.push("Commit not signed off");
+            result = false;
+        }
```

Comment:
I feel that this might be simpler if we do two things: introduce a method recording the message and whether there were blocking problems, and reordering the conditional arms:

```javascript
private async blocker(message: string): Promise<void> {
    this.addComment(message);
    this.result = false;
}
[...]
    if (!match) {
        blocker("Commit not signed off");
    } else if (match[1].length) {
        blocker(`Leading whitespace in sign off: ${line}`);
    }
```